pkg/agent/discovery_agent.go

```diff
@@ -0,0 +1,255 @@
+package agent
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"github.com/hashicorp/go-version"
+	configv1 "github.com/openshift/api/config/v1"
+	hyperv1beta1 "github.com/openshift/hypershift/api/hypershift/v1beta1"
+	discoveryv1 "github.com/stolostron/discovery/api/v1"
+	"github.com/stolostron/hypershift-addon-operator/pkg/util"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+type DiscoveryAgent struct {
+	hubClient   client.Client
+	spokeClient client.Client
+	clusterName string
+	log         logr.Logger
+}
+
+var DiscoveryPredicateFunctions = predicate.Funcs{
```

Comment:
Please add comment for the variable and what is suppose to be the use of GenericFunc